### PR TITLE
[rllib] Documentation for I/O API and multi-agent support / cleanup

### PR DESCRIPTION
**Pull request by ericl**
## What do these changes do?

Adds multi-agent support, some cleanup of unsupported features. I think that in the initial release, we will _only_ support SampleBatch input/output types. This has the following advantages:
- works with both on-policy and off-policy algorithms
- higher performance
- much simpler to implement

The disadvantages are that:
- you can't change the observation preprocessing as the experiences are already postprocessed
- more inconvenient to ingest real data, as you first have to convert it to SampleBatch format with JsonWriter

I think we can probably offset the disadvantages with examples on how to use JsonWriter, and punt the ability to process non-batch input for now. The user will have to some some preprocessing of their data anyways before it is ready for ingest.

## Related issue number

Closes https://github.com/ray-project/ray/issues/3363

cc @richardliaw @joneswong

### CODE REVIEW

**Comment by ericl** [comment]
Jenkins retest this please

**Comment by ericl** [comment]
Jenkins retest this please

File: doc/source/rllib-offline.rst [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+You can configure experience output for an agent using the following options:
+
+.. literalinclude:: ../../python/ray/rllib/agents/agent.py
+   :language: python
+   :start-after: __sphinx_doc_output_begin__
+   :end-before: __sphinx_doc_output_end__
+
+The interface for a custom output writer is as follows:
+
+.. literalinclude:: ../../python/ray/rllib/offline/output_writer.py
```
[the remark this answers is not in the document]
Comment (ericl) [reply doc/source/rllib-offline.rst]: Fixed

File: python/ray/rllib/evaluation/policy_evaluator.py [hunk 2]
```diff
@@ -309,8 +307,6 @@ def make_env(vector_index):
                 "Requested 'simulation' input evaluation method: "
                 "will discard all sampler outputs and keep only metrics.")
             sample_async = True
-        elif input_evaluation_method == "counterfactual":
```
[the remark this answers is not in the document]
Comment (ericl) [reply python/ray/rllib/evaluation/policy_evaluator.py]: @joneswong didn't find it super useful in his use cases. We can consider adding it later, but for now it's not implemented.

**Review by richardliaw — APPROVED** [review]

**Pull request merged**